src/testing/flow_solver.cpp

```diff
@@ -1,35 +1,324 @@
 , initial_condition_function(InitialConditionFactory<dim,double>::create_InitialConditionFunction(parameters_input, nstate))
 , all_param(*parameters_input)
 , flow_solver_param(all_param.flow_solver_param)
 , ode_param(all_param.ode_solver_param)
 , poly_degree(all_param.grid_refinement_study_param.poly_degree)
 , final_time(flow_solver_param.final_time)
+, input_parameters_file_reference_copy_filename(flow_solver_param.restart_files_directory_name + std::string("/") + std::string("input_copy.prm"))
 , dg(DGFactory<dim,double>::create_discontinuous_galerkin(&all_param, poly_degree, flow_solver_case->generate_grid()))
 , ode_solver(ODE::ODESolverFactory<dim, double>::create_ODESolver(dg))
 {
     dg->allocate_system();
     flow_solver_case->display_flow_solver_setup();
-    pcout << "Initializing solution with initial condition function..." << std::flush;
+    
     dealii::LinearAlgebra::distributed::Vector<double> solution_no_ghost;
     solution_no_ghost.reinit(dg->locally_owned_dofs, MPI_COMM_WORLD);
-    dealii::VectorTools::interpolate(dg->dof_handler, *initial_condition_function, solution_no_ghost);
-    dg->solution = solution_no_ghost;
+
+    if(flow_solver_param.restart_computation_from_file == true) {
+        if(dim == 1) {
+            pcout << "Error: restart_computation_from_file is not possible for 1D. Set to false." << std::endl;
+            std::abort();
+        }
+#if PHILIP_DIM>1
+        if (flow_solver_param.steady_state == true) {
+            pcout << "Error: Restart capability has not been fully implemented / tested for steady state computations." << std::endl;
+            std::abort();
+        }
+
+        // Initialize solution from restart file
+        pcout << "Initializing solution from restart file..." << std::flush;
+        const std::string restart_filename_without_extension = get_restart_filename_without_extension(flow_solver_param.restart_file_index);
+        dg->triangulation->load(flow_solver_param.restart_files_directory_name + std::string("/") + restart_filename_without_extension);
+        
+        // Note: Future development with hp-capabilities, see section "Note on usage with DoFHandler with hp-capabilities"
+        // ----- Ref: https://www.dealii.org/current/doxygen/deal.II/classparallel_1_1distributed_1_1SolutionTransfer.html
+        dealii::parallel::distributed::SolutionTransfer<dim, dealii::LinearAlgebra::distributed::Vector<double>, dealii::DoFHandler<dim>> solution_transfer(dg->dof_handler);
+        solution_transfer.deserialize(solution_no_ghost);
+#endif
+    } else {
+        // Initialize solution from initial_condition_function
+        pcout << "Initializing solution with initial condition function... " << std::flush;
+        dealii::VectorTools::interpolate(dg->dof_handler, *initial_condition_function, solution_no_ghost);
+    }
+    dg->solution = solution_no_ghost; //< assignment
+    // dg->solution.update_ghost_values(); // -- should this be done ?
```

Comment:
Maybe. Safer to do it and inexpensive. DG should automatically update the ghost values if you evaluate the residual. However, if you do something else with the solution that requires ghost values, then this is needed.

Reply:
Thanks for the explanation, I've addressed this in the most recent commit